islandora.tokens.inc

```diff
@@ -0,0 +1,124 @@
+  // Deprecated in favour if hyphenated version.
+  $node['media_thumbnail_image:url'] = [
+    'name' => t('Media: Thumbnail Image URL.'),
+    'description' => t('Deprecated: URL of Thumbnail Image associated with Islandora Object via Media.'),
+  ];
+
+  // Deprecated in favour if hyphenated version.
+  $node['media_thumbnail_image:alt'] = [
+    'name' => t('Alternative text for Media: Thumbnail Image.'),
+    'description' => t('Deprecated: Alternative text for Thumbnail Image associated with Islandora Object via Media.'),
+  ];
+
+
+  return [
+    'types' => ['islandoratokens' => $type],
+    'tokens' => ['islandoratokens' => $node],
+  ];
+}
+
+/**
+ * Implements hook_tokens().
+ */
+function islandora_tokens($type, $tokens, array $data, array $options, BubbleableMetadata $bubbleable_metadata) {
+  $replacements = [];
+  if ($type == 'islandoratokens' && !empty($data['node'])) {
+    if (!is_array($tokens) || empty($tokens)) {
+      \Drupal::logger('controlled_access_terms')
+        ->alert('Tokens not correct format: @tokens', [
+          '@tokens' => print_r($tokens, 1),
+        ]);
+      return;
+    }
+    foreach ($tokens as $name => $original) {
+      switch ($name) {
+        case 'media-thumbnail-image:url':
+        case 'media_thumbnail_image:url':
+          $term = $islandoraUtils->getTermForUri('http://pcdm.org/use#ThumbnailImage');
+          $media = $islandoraUtils->getMediaWithTerm($data['node'], $term);
+          // Is there media?
+          // @todo: is this single or multiple?
+          if ($media) {
+            $file = controlled_access_terms_image_from_media($media);
+            if (!empty($file)) {
+              $url = $file->url();
+              $replacements[$original] = $url;
+            }
+          }
+          break;
+
+        case 'media-thumbnail-image:alt':
+        case 'media_thumbnail_image:alt':
+          $alt = '';
+          $islandoraUtils = \Drupal::service('islandora.utils');
+          $term = $islandoraUtils->getTermForUri('http://pcdm.org/use#ThumbnailImage');
+          $media = $islandoraUtils->getMediaWithTerm($data['node'], $term);
+          // Is there media?
+          // @todo: is this single or multiple?
+          if ($media) {
+            // Is the media an image?
+            if (isset($media->field_media_image)) {
+              $alt = $media->field_media_image[0]->alt;
+            }
+          }
+          // @todo: get alt from original or service file, if thumbnail alt is empty.
+          $replacements[$original] = $alt;
+          break;
+      }
+    }
+  }
+  return $replacements;
+}
+
+/**
+ * Get File referenced by given Media.
+ *
+ * @todo: Add to IslandoraUtils?
+ * @todo: Had to remove typehint MediaInterface from function.
+ *
+ * @param \Drupal\media\MediaInterface $media
+ *   The Media to operate on.
+ *
+ * @return mixed
+ *   \Drupal\file\FileInterface on success, NULL otherwise.
+ */
+function controlled_access_terms_image_from_media($media) {
```

Comment:
Probably can ditch this for the MediaSourceService, but if not, you'd wanna rename it :)